README.md
<img src="http://opendatadpc.maps.arcgis.com/sharing/rest/content/items/5c8ef7516b5b4bb19f61037b4cd69015/data" alt="COVID-19" data-canonical-src="http://opendatadpc.maps.arcgis.com/sharing/rest/content/items/5c8ef7516b5b4bb19f61037b4cd69015/data" width="400" />

[Italiano](README.md) - [English](README_EN.md)<br><br>
<<<<<<< HEAD
=======
<<<<<<< HEAD
>>>>>>> master

# Dati COVID-19 Italia

[![GitHub license](https://img.shields.io/badge/License-Creative%20Commons%20Attribution%204.0%20International-blue)](https://github.com/pcm-dpc/COVID-19/blob/master/LICENSE)
[![GitHub commit](https://img.shields.io/github/last-commit/pcm-dpc/COVID-19)](https://github.com/pcm-dpc/COVID-19/commits/master)
 
<<<<<<< HEAD
=======
## Avviso modifiche 
[https://github.com/pcm-dpc/COVID-19/issues/625](https://github.com/pcm-dpc/COVID-19/issues/625)<br><br>
>>>>>>> 44dde7ca11b83b24840315d553018c817698452f

[Sito del Dipartimento della Protezione Civile - Emergenza Coronavirus: la risposta nazionale](http://www.protezionecivile.it/attivita-rischi/rischio-sanitario/emergenze/coronavirus)
=======

# Dati COVID-19 Italia
>>>>>>> 59e796d5adc083cd3124f2be8a3d0bf13e2782fb

[![GitHub license](https://img.shields.io/badge/License-Creative%20Commons%20Attribution%204.0%20International-blue)](https://github.com/pcm-dpc/COVID-19/blob/master/LICENSE)
[![GitHub commit](https://img.shields.io/github/last-commit/pcm-dpc/COVID-19)](https://github.com/pcm-dpc/COVID-19/commits/master)
 
[Sito del Dipartimento della Protezione Civile - Emergenza Coronavirus: la risposta nazionale](http://www.protezionecivile.it/attivita-rischi/rischio-sanitario/emergenze/coronavirus)

Il 31 gennaio 2020, il Consiglio dei Ministri dichiara lo stato di emergenza, per la durata di sei mesi, in conseguenza del rischio sanitario connesso all'infezione da Coronavirus.

Al Capo del Dipartimento della Protezione Civile, Angelo Borrelli, è affidato il coordinamento degli interventi necessari a fronteggiare l'emergenza sul territorio nazionale.  
  
Le principali azioni coordinate dal Capo del Dipartimento sono volte al soccorso e all'assistenza della popolazione eventualmente interessata dal contagio, al potenziamento dei controlli nelle aree aeroportuali e portuali, in continuità con le misure urgenti già adottate dal Ministero della salute, al rientro in Italia dei cittadini che si trovano nei Paesi a rischio e al rimpatrio dei cittadini stranieri nei Paesi di origine esposti al rischio.

Per informare i cittadini e mettere a disposizione i dati raccolti, utili ai soli fini comunicativi e di informazione, il Dipartimento della Protezione Civile ha elaborato un cruscotto geografico interattivo raggiungibile agli indirizzi  [http://arcg.is/C1unv](http://arcg.is/C1unv) (versione desktop) e [http://arcg.is/081a51](http://arcg.is/081a51) (versione mobile) e mette a disposizione, con licenza CC-BY-4.0, le seguenti informazioni aggiornate quotidianamente alle 18:30 (successivamente la conferenza stampa del Capo Dipartimento):

- Dati Andamento nazionale
- Dati json
- Dati regioni
- Dati province
- Schede riepilogative
- Aree
- Note
- Dati contratti DPC forniture

## Avvisi

[Avvisi sui dati andamento COVID-19 Italia](avvisi.md)<br>

## Struttura del repository
```
COVID-19/
│
├── aree/
│   ├── geojson
│   │   ├── dpc-covid-19-ita-aree-comuni.geojson
│   │   ├── dpc-covid19-ita-aree.geojson
│   ├── shp
│   │   ├── dpc-covid19-ita-aree-comuni.dbf
│   │   ├── dpc-covid19-ita-aree-comuni.prj
│   │   ├── dpc-covid19-ita-aree-comuni.shp
│   │   ├── dpc-covid19-ita-aree-comuni.shx
│   │   ├── dpc-covid19-ita-aree.dbf
│   │   ├── dpc-covid19-ita-aree.prj
│   │   ├── dpc-covid19-ita-aree.shp
│   │   ├── dpc-covid19-ita-aree.shx
├── dati-andamento-nazionale/
│   ├── dpc-covid19-ita-andamento-nazionale-*.csv
│   ├── dpc-covid19-ita-andamento-nazionale-latest.csv
│   ├── dpc-covid19-ita-andamento-nazionale.csv
├── dati-contratti-dpc-forniture/
│   ├── dpc-covid19-dati-contratti-dpc-forniture.csv
│   ├── dpc-covid19-dati-pagamenti-contratti-dpc-forniture.csv
│   ├── dati-json
│   │   ├── dpc-covid19-dati-contratti-dpc-forniture.csv
│   │   ├── dpc-covid19-dati-pagamenti-contratti-dpc-forniture.csv
│   ├── file-atti-negoziali
│   │   ├── dpc-contratto-covid19-*.pdf
├── dati-json/
│   ├── dpc-covid19-ita-andamento-nazionale-latest.json
│   ├── dpc-covid19-ita-andamento-nazionale.json
│   ├── dpc-covid19-ita-note-en.json
│   ├── dpc-covid19-ita-note-it.json
│   ├── dpc-covid19-ita-province-latest.json
│   ├── dpc-covid19-ita-province.json
│   ├── dpc-covid19-ita-regioni-latest.json
│   ├── dpc-covid19-ita-regioni.json
├── dati-province/
│   ├── dpc-covid19-ita-province-*.csv
│   ├── dpc-covid19-ita-province-latest.csv
│   ├── dpc-covid19-ita-province.csv
├── dati-regioni/
│   ├── dpc-covid19-ita-regioni-*.csv
│   ├── dpc-covid19-ita-regioni-latest.csv
│   ├── dpc-covid19-ita-regioni.csv
<<<<<<< HEAD
=======
<<<<<<< HEAD
=======
├── data-legacy
>>>>>>> 59e796d5adc083cd3124f2be8a3d0bf13e2782fb
>>>>>>> master
├── note/
│   ├── dpc-covid19-ita-note-en.csv
│   ├── dpc-covid19-ita-note-it.csv
├── schede-riepilogative/
│   ├── province
│   │   ├── dpc-covid19-ita-scheda-province-*.pdf
│   ├── regioni
│   │   ├── dpc-covid19-ita-scheda-regioni-*.pdf
```

## Aggiornamento dei dati

- Dati andamento COVID-19 Italia: ogni giorno alle 18:00<br>
- Dati contratti DPC COVID-19 di fornitura: continua (ogni volta che vengono effettuate operazioni sui contratti)
<<<<<<< HEAD

## Formato dei dati

- [Dati andamento COVID-19 Italia](dati-andamento-covid19-italia.md)<br>
- [Dati contratti DPC COVID-19 di fornitura](dati-contratti-dpc-covid19-fornitura.md)
- [Dati aree misure restrittive COVID19](dati_aree-nazionali-subregionali-misure-restrittive-covid19.md)
=======

## Formato dei dati

- [Dati andamento COVID-19 Italia](dati-andamento-covid19-italia.md)<br>
- [Dati contratti DPC COVID-19 di fornitura](dati-contratti-dpc-covid19-fornitura.md)
- [Dati aree misure restrittive COVID19](dati-aree-covid19.md)
>>>>>>> 59e796d5adc083cd3124f2be8a3d0bf13e2782fb

## Licenza

[CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/deed.it) - [Visualizza licenza](https://github.com/pcm-dpc/COVID-19/blob/master/LICENSE)


CHANGELOG.md
# Changelog

Tutte le modifiche al progetto, nuove funzionalità e informazioni sono documentate in questo file

## 2020-03-08

- Modifica: "dati-andamento-nazionale" riportando i totali dei dati delle Regioni
- Rimossa: directory "shape-aree-contenimento" e creata "aree" con "shp" e "geojson"

## 2020-03-10

- Aggiornate le Aree secondo il nuovo DPCM 9 Marzo 2020

## 2020-03-11

- Modifica: "Bolzano" e "Trento" rinominati in "P.A. Bolzano" e "P.A. Trento" in "dati-regioni e "dati-province" ("denominazione_regione")
- Modifica: Friuli V. G. rinominata in "dati-regioni e "dati-province" ("denominazione_regione") in "Friuli Venezia Giulia"

## 2020-03-12

- Aggiunto: Codice di Condotta in italiano (CODE_OF_CONDUCT.md) e in inglese (CODE_OF_CONDUCT_EN.md)
- Aggiunto: Changelog in italiano (CHANGELOG.md) e in inglese (CHANGELOG_EN.md)
- Aggiornate le Aree secondo il nuovo DPCM 11 Marzo 2020

## 2020-03-19

- Aggiunti ultimi dati (latest) csv, nelle rispettive directory, per andamento nazionale (dpc-covid19-ita-andamento-nazionale-latest.csv), regioni (dpc-covid19-ita-regioni-latest.csv) e province (dpc-covid19-ita-province-latest.csv)
- Aggiunti ultimi dati (latest) json, nella directory dati-json, per andamento nazionale (dpc-covid19-ita-andamento-nazionale-latest.json), regioni (dpc-covid19-ita-regioni-latest.json) e province (dpc-covid19-ita-province-latest.json)

## 2020-03-25

- Data nel formato ISO8601 UTC
- Aggiunta: "Note" in "dati-regioni", "dati-province" e "dati-andamento-nazionale"
- Aggiunta: dataset "note"

## 2020-03-30

- Modifica: "totale_attualmente_positivi" rinominato in "totale_positivi" (ricoverati_con_sintomi + terapia_intensiva + isolamento domiciliare) in "dati_regioni" e "dati_andamento_nazionale"
- Modifica: "nuovi_attualmente_positivi" rinominato in "variazione_totale_positivi" (totale_attualmente positivi giorno corrente - totale_attualmente positivi giorno precedente) in "dati_regioni" e "dati_andamento_nazionale"
- Aggiunta: "nuovi_positivi" (totale_casi giorno corrente - totale_casi giorno precedente) in "dati_regioni" e "dati_andamento_nazionale"
- Modifica: Regione "Emilia Romagna" rinominato in "Emilia-Romagna" in "dati-regioni" e "dati-province" ("denominazione_regione")

## 2020-04-20

- Aggiunta: "casi_testati", totale dei soggetti sottoposti al test dal 19/04/2020

## 2020-04-21

<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
- Aggiunta: nuovo shape file con misure di contenimento a livello subregionale (province, comuni e frazioni).
=======
=======
>>>>>>> master
- Aggiunta: nuovo shape file con misure di contenimento a livello subregionale (province, comuni e frazioni) con rlativi metadata.
>>>>>>> 44dde7ca11b83b24840315d553018c817698452f
=======
- Aggiunta: nuovo shape file con misure di contenimento a livello subregionale (province, comuni e frazioni) con rlativi metadata.
>>>>>>> master

## 2020-04-22

- Aggiunta: nuovo dataset contratti dpc forniture (contratti e pagamenti) in formato csv e json.  [link alla dashboard](https://app.powerbi.com/view?r=eyJrIjoiNzA1YmQ0NDctYzY3ZC00ZTlkLWI1MTQtMThmNTEwNWY3NjM4IiwidCI6IjlhMDZhOTA3LTA2OTUtNDA0YS05NmY4LTRhMWU5YTJmYjQxZCIsImMiOjl9) - [link al dataset ](https://github.com/pcm-dpc/COVID-19/tree/master/dati-contratti-dpc-forniture)

## 2020-04-23

- Aggiunta: metadati del nuovo dataset contratti dpc forniture (contratti e pagamenti) in formato DCAT-AP-IT
- Integrato README con specifiche per il data model dei dataset aree (nazionali e subregionali)

## 2020-04-27

<<<<<<< HEAD
<<<<<<< HEAD
- Aggiunta: aggiornamento al26/4/2020 shape file con misure di contenimento a livello subregionale (province, comuni e frazioni).
=======
- Aggiunta: aggiornamento al 26/4/2020 shape file e relativi metadata con misure di contenimento a livello subregionale (province, comuni e frazioni).

## 2020-05-04

- Aggiunta: aggiornamento al 4/5/2020 shape file e relativi metadati con misure di contenimento a livello nazionale a seguito del DPCM 26/4/2020.
<<<<<<< HEAD
>>>>>>> master
=======
>>>>>>> master

<<<<<<< HEAD
## Prossimi aggiornamenti

- API Rest (Json)
- GraphQL

## Ipotesi modifiche
=======
- Aggiunta: aggiornamento al 26/4/2020 shape file e relativi metadata con misure di contenimento a livello subregionale (province, comuni e frazioni).

## 2020-05-04

- Aggiunta: aggiornamento al 4/5/2020 shape file e relativi metadati con misure di contenimento a livello nazionale a seguito del DPCM 26/4/2020.

## 2020-05-05

- API Rest (Json) private beta
- GraphQL private beta

## 2020-05-06

- Aggiunta: aggiornamento al 03/05/2020 shape file e relativi metadata con misure di contenimento a livello subregionale (province, comuni e frazioni).

## Prossimi aggiornamenti
>>>>>>> 44dde7ca11b83b24840315d553018c817698452f

- Cambio codice_regione P.A. Bolzano e P.A. Trento: da 04 a 21 per P.A. Bolzano e da 04 a 22 per P.A. Trento (codice provincia)
=======
## 2020-05-06

- Aggiunta: aggiornamento al 03/05/2020 shape file e relativi metadata con misure di contenimento a livello subregionale (province, comuni e frazioni).

## 2020-05-16

- Cambio codice_regione P.A. Bolzano e P.A. Trento: da 04 a 21 per P.A. Bolzano e da 04 a 22 per P.A. Trento (codice provincia)

## 2020-05-18

- Aggiunta: aggiornamento al 18/5/2020 shape file e relativi metadati con misure di contenimento a livello nazionale a seguito del DPCM 17/5/2020.

## Prossimi aggiornamenti

- API Rest (Json) public beta
- GraphQL public beta
>>>>>>> 59e796d5adc083cd3124f2be8a3d0bf13e2782fb


frcovid19_old.py
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
import math as m 
import scipy as sp 
import csv

def confidence_ellipse(x,y,ax,width_coeff=6,**kwargs):
    """
    from https://stackoverflow.com/questions/20126061/creating-a-confidence-ellipses-in-a-sccatterplot-using-matplotlib
    """

    from matplotlib.patches import Ellipse
    
    cov = np.cov(x, y)
    lambda_, v = np.linalg.eig(cov)
    lambda_ = np.sqrt(lambda_)  
    ell = Ellipse(xy=(np.mean(x), np.mean(y)),width=lambda_[0]*width_coeff, height=lambda_[1]*width_coeff,angle=np.rad2deg(np.arccos(v[0, 0])),**kwargs)

    return ax.add_patch(ell)

def last_update(dates,language):
    """
    A very short function that prints when the analysis was latest updated.

    INPUT:
    dates = the list with the dates in the data dictionary
    language = ITA or ENG
    """
    if language =='ITA':
        print("Ultimo aggiornamento il {}".format(dates[-1]))
    elif language == 'ENG':
        print("Last updated on {}".format(dates[-1]))

def region_names():
    """
    reminder of region names
    """
    regnames = ['Abruzzo', 'Basilicata', 'P.A. Bolzano', 'Calabria', 'Campania','Emilia-Romagna','Friuli Venezia Giulia','Lazio','Liguria','Lombardia','Marche','Molise','Piemonte','Puglia','Sardegna','Sicilia','Toscana','P.A. Trento']
    print('Seleziona la regione tra le seguenti. Attenzione agli spazi, trattini, ecc./These are the names of the regions in Itaian.')
    print(regnames)


def data_loader_generic(filename):
    """
    generic csv data loader. Takes a csv file, 
    and gives out a dict in which the keys are 
    defined by the first row of the csv file.
    All elements are string.

    INPUT:

    filename = PATH string of the csv file

    OUTPUT: 
    
    out = a dictionary
    keys = the keys of out
    """
        #initialize reader
    content = []
    #open the file and load the contents into it
    with open(filename) as datafile:
        rawcontents = csv.reader(datafile, delimiter=',')
        for row in rawcontents:
            content.append(row)

    keys = content[0]
    out={}
    for i in range(0,len(keys)):
        out[keys[i]] = [];
        for j in range(1,len(content)):
            out[keys[i]].append(content[j][i])
    
    return keys, out

def data_loader_nazionali(filename):
    """
    This function loads the data from the Protezione Civile and 
    returns a dictionary in which they are all divided by field. 
    The dates are in MM-DD format. This version leaves the least
    control to the user.

    INPUT:
    filname = PATH string of the Protezione Civile .csv file

    OUTPUT
    out = a dictionary containing the data. The Keys for this dict are the fields in the CSV file
    """
    keys, content = data_loader_generic(filename)
    out={}
    for key in keys:
        out[key] = []
        for element in content[key]:
            if key == 'data':
                #dates get separated from the 'T18:00:00' string, the format gets switched to DD/MM
                giorno = element[8:10]
                mese = element[5:7]
                out[key].append((giorno + '/'+ mese))
            elif key not in ['note_en','note_it','stato']:
                out[key].append(int(element))

    for key in ['note_en','note_it','stato']:
        out.pop(key)

    return out

def data_loader_region(filename,region_name):
    """
    This function loads the file where the data by region is stored. 
    The file structure is a little different and contains a number identifier by region, that we can use.
    region_handle is necessary. Unfortunately, our life is made difficult by the fact 
    that Trento and Bolzano have different names, but the same region code.
    For now, these two provinces will be considered separately.
    """
    
    keys, content = data_loader_generic(filename)
    out = {}
    for key in keys:
        out[key] = []
        for j in range(0,len(content[key])):
            if content['denominazione_regione'][j] == region_name:
                if key == 'data':
                    giorno = content[key][j][8:10]
                    mese = content[key][j][5:7]
                    out[key].append((giorno + '/'+ mese))
                elif key not in ['stato','denominazione_regione','lat','long','note_it','note_en']:
                    out[key].append(int((content[key][j])))
    
    for key in ['stato','denominazione_regione','lat','long','note_it','note_en']:
        out.pop(key)
    
    return out

def moving_average(x,N=1,mode = 'same'):
    ''' This is a function that makes a moving average of a 1d vector using numpy.convolve
    INPUTS: 
    x: one-dimensional vector
    N: dimension of the moving average
    mode: the mode of numpy.convolve, that determines the behaviour at the edges of the moving average. Valid values are full, same and valid. Check the doc for numpy.convolve for details.
    OUTPUT:
    out: the result of the averaging operation'''
    
    import numpy as np

    averaging_window = ([1]*N)
    averaging_window = [i/N for i in averaging_window]

    return np.convolve(x,averaging_window,mode)


def data_loader_nazionali_OLD(filename):
    """
    DEPRECATED
    This function loads the data from the Protezione Civile and 
    returns a dictionary in which they are all divided by field. 
    The dates are in MM-DD format. This version leaves the least
    control to the user.

    INPUT:
    filname = PATH string of the Protezione Civile .csv file

    OUTPUT
    out = a dictionary containing the data. The Keys for this dict are the fields in the CSV file
    """
    #initialize reader
    content = []
    #open the file and load the contents into it
    with open(filename) as datafile:
        rawcontents = csv.reader(datafile, delimiter=',')
        for row in rawcontents:
            content.append(row)
    
    #define the dictionary keys from the reader
    keys = content[0];
    #initialize output
    out={}
    for i in range(0,len(keys)):
        out[keys[i]] = [];
    #load the data into the dictionary
    for i in range(0,len(keys)):
        for j in range(1,len(content)):
            if 2<i<13:
                 #numerical data is passed as numbers
                out[keys[i]].append(int(content[j][i]))
            elif keys[i] == 'data':
                #dates get separated from the 'T18:00:00' string
                giorno = content[j][i][8:10]
                mese = content[j][i][5:7]
                out[keys[i]].append((giorno + '/'+ mese))
                #out[keys[i]].append(content[j][i][5:10])  
            else:
                #The rest is passed as it is
                out[keys[i]].append(content[j][i])
    
    #return the data
    return out




frcovid19.py
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
import math as m 
import scipy as sp 
import csv
from matplotlib import cm


class DatiCOVID():
    #thecolormap = cm.get_cmap('gnuplot2_r',4*complete_weeks+3)
    #week_no = []
    def __init__(self,args):
        self.day = [x for x in range(1,len(args['data']))]
        self.data = args['data']
        self.ricoverati_con_sintomi = args['ricoverati_con_sintomi']
        self.terapia_intensiva = args['terapia_intensiva']
        self.totale_ospedalizzati = args['totale_ospedalizzati']
        self.isolamento_domiciliare = args['isolamento_domiciliare']
        self.totale_positivi = args['totale_positivi']
        self.variazione_totale_positivi = args['variazione_totale_positivi']
        self.nuovi_positivi = args['nuovi_positivi']
        self.dimessi_guariti = args['dimessi_guariti']
        self.deceduti = args['deceduti']
        self.totale_casi = args['totale_casi']
        self.tamponi = args['tamponi']
        self.casi_testati = args['casi_testati']
        self.settimana = ['Settimana 1']
        for i in range(1,len(self.data)):
           self.settimana.append('Settimana ' + str(m.ceil((i+1)/7)))
        #self.settimana = settimana
        self.tamponi_giornalieri = [self.tamponi[0]]
        for i in range(1,len(self.tamponi)):
            self.tamponi_giornalieri.append(self.tamponi[i]-self.tamponi[i-1])

    def dati_settimanali(self,dataset,week):
        week_string = 'Settimana '+str(week)
        dati = 0
        dati = [x for (t,x) in zip(self.settimana,dataset) if t == week_string]
        return dati

    def totale_settimanale(self,dataset,week):
        week_string = 'Settimana '+str(week)
        totale = sum([x for (t,x) in zip(self.settimana,dataset) if t == week_string])
        return totale

def confidence_ellipse(x,y,ax,width_coeff=6,**kwargs):
    """
    from https://stackoverflow.com/questions/20126061/creating-a-confidence-ellipses-in-a-sccatterplot-using-matplotlib
    """

    from matplotlib.patches import Ellipse
    
    cov = np.cov(x, y)
    lambda_, v = np.linalg.eig(cov)
    lambda_ = np.sqrt(lambda_)  
    ell = Ellipse(xy=(np.mean(x), np.mean(y)),width=lambda_[0]*width_coeff, height=lambda_[1]*width_coeff,angle=np.rad2deg(np.arccos(v[0, 0])),**kwargs)

    return ax.add_patch(ell)

def last_update(dates,language):
    """
    A very short function that prints when the analysis was latest updated.

    INPUT:
    dates = the list with the dates in the data dictionary
    language = ITA or ENG
    """
    if language =='ITA':
        print("Ultimo aggiornamento il {}".format(dates[-1]))
    elif language == 'ENG':
        print("Last updated on {}".format(dates[-1]))

def region_names():
    """
    reminder of region names
    """
    regnames = ['Abruzzo', 'Basilicata', 'P.A. Bolzano', 'Calabria', 'Campania','Emilia-Romagna','Friuli Venezia Giulia','Lazio','Liguria','Lombardia','Marche','Molise','Piemonte','Puglia','Sardegna','Sicilia','Toscana','P.A. Trento']
    print('Seleziona la regione tra le seguenti. Attenzione agli spazi, trattini, ecc./These are the names of the regions in Itaian.')
    print(regnames)


def data_loader_generic(filename):
    """
    generic csv data loader. Takes a csv file, 
    and gives out a dict in which the keys are 
    defined by the first row of the csv file.
    All elements are string.

    INPUT:

    filename = PATH string of the csv file

    OUTPUT: 
    
    out = a dictionary
    keys = the keys of out
    """
        #initialize reader
    content = []
    #open the file and load the contents into it
    with open(filename) as datafile:
        rawcontents = csv.reader(datafile, delimiter=',')
        for row in rawcontents:
            content.append(row)

    keys = content[0]
    out={}
    for i in range(0,len(keys)):
        out[keys[i]] = [];
        for j in range(1,len(content)):
            if content[j][0][0] == '2':
                out[keys[i]].append(content[j][i])
            else:
                continue
    
    return keys, out

def data_loader_nazionali(filename):
    """
    This function loads the data from the Protezione Civile and 
    returns a dictionary in which they are all divided by field. 
    The dates are in MM-DD format. This version leaves the least
    control to the user.

    INPUT:
    filname = PATH string of the Protezione Civile .csv file

    OUTPUT
    out = a dictionary containing the data. The Keys for this dict are the fields in the CSV file
    """
    keys, content = data_loader_generic(filename)
    out={}
    for key in keys:
        out[key] = []
        for element in content[key]:
            if key == 'data':
                #dates get separated from the 'T18:00:00' string, the format gets switched to DD/MM
                giorno = element[8:10]
                mese = element[5:7]
                out[key].append((giorno + '/'+ mese))
            elif key not in ['note_en','note_it','stato']:
                if element == '':
                    out[key].append(0) # fix empty entries
                else:
                    out[key].append(int(element))

    for key in ['note_en','note_it','stato']:
        out.pop(key)

    return out

def data_loader_region(filename,region_name):
    """
    This function loads the file where the data by region is stored. 
    The file structure is a little different and contains a number identifier by region, that we can use.
    region_handle is necessary. Unfortunately, our life is made difficult by the fact 
    that Trento and Bolzano have different names, but the same region code.
    For now, these two provinces will be considered separately.
    """
    
    keys, content = data_loader_generic(filename)
    out = {}
    for key in keys:
        out[key] = []
        for j in range(0,len(content[key])):
            if content['denominazione_regione'][j] == region_name:
                if key == 'data':
                    giorno = content[key][j][8:10]
                    mese = content[key][j][5:7]
                    out[key].append((giorno + '/'+ mese))
                elif key not in ['stato','denominazione_regione','lat','long','note_it','note_en']:
                    if content[key][j] == '':
                        out[key].append(0) # fix empty entries
                    else:
                        out[key].append(int(content[key][j]))
    
    for key in ['stato','denominazione_regione','lat','long','note_it','note_en','codice_regione']:
        out.pop(key)
    
    return out

def moving_average(x,N=1,mode = 'same'):
    ''' This is a function that makes a moving average of a 1d vector using numpy.convolve
    INPUTS: 
    x: one-dimensional vector
    N: dimension of the moving average
    mode: the mode of numpy.convolve, that determines the behaviour at the edges of the moving average. Valid values are full, same and valid. Check the doc for numpy.convolve for details.
    OUTPUT:
    out: the result of the averaging operation'''
    
    import numpy as np

    averaging_window = ([1]*N)
    averaging_window = [i/N for i in averaging_window]

    return np.convolve(x,averaging_window,mode)


def data_loader_nazionali_OLD(filename):
    """
    DEPRECATED
    This function loads the data from the Protezione Civile and 
    returns a dictionary in which they are all divided by field. 
    The dates are in MM-DD format. This version leaves the least
    control to the user.

    INPUT:
    filname = PATH string of the Protezione Civile .csv file

    OUTPUT
    out = a dictionary containing the data. The Keys for this dict are the fields in the CSV file
    """
    #initialize reader
    content = []
    #open the file and load the contents into it
    with open(filename) as datafile:
        rawcontents = csv.reader(datafile, delimiter=',')
        for row in rawcontents:
            content.append(row)
    
    #define the dictionary keys from the reader
    keys = content[0];
    #initialize output
    out={}
    for i in range(0,len(keys)):
        out[keys[i]] = [];
    #load the data into the dictionary
    for i in range(0,len(keys)):
        for j in range(1,len(content)):
            if 2<i<13:
                 #numerical data is passed as numbers
                out[keys[i]].append(int(content[j][i]))
            elif keys[i] == 'data':
                #dates get separated from the 'T18:00:00' string
                giorno = content[j][i][8:10]
                mese = content[j][i][5:7]
                out[keys[i]].append((giorno + '/'+ mese))
                #out[keys[i]].append(content[j][i][5:10])  
            else:
                #The rest is passed as it is
                out[keys[i]].append(content[j][i])
    
    #return the data
    return out


